server/server/src/billing/billing_service.rs

```diff
@@ -0,0 +1,500 @@
+                )
+                .await
+                .map_err(|e| {
+                    internal!(
+                        "While trying to get the expanded invoice, an error was encountered: {:?}",
+                        e
+                    )
+                })?;
+
+                let subscription_period_end = match invoice.subscription {
+                    None => {
+                        return Err(internal!(
+                            "There should be a subscription tied to this invoice: {:?}",
+                            invoice
+                        ))
+                    }
+                    Some(Expandable::Id(_)) => {
+                        return Err(internal!(
+                            "The subscription should be expanded in this invoice: {:?}",
+                            invoice
+                        ))
+                    }
+                    Some(Expandable::Object(subscription)) => subscription.current_period_end,
+                };
+
+                let customer_id = match invoice.customer.ok_or_else(|| {
+                    ClientError(StripeWebhookError::InvalidBody(
+                        "Cannot retrieve the customer_id.".to_string(),
+                    ))
+                })? {
+                    Expandable::Id(id) => id.to_string(),
+                    Expandable::Object(customer) => customer.id.to_string(),
+                };
+
+                let (public_key, mut user_info) =
+                    get_public_key_and_stripe_user_info(&event, &mut con, &customer_id).await?;
+                let pos = get_active_subscription_index(&user_info.subscriptions)?;
+
+                info!(
+                    "Stripe successfully renewed a user's subscription. The user's subscription period_end is being changed. public_key: {}",
+                    keys::stringify_public_key(&public_key)
+                );
+
+                user_info.subscriptions[pos].period_end = subscription_period_end as u64;
+
+                con.json_set(stripe_user_info(&public_key), &user_info)
+                    .await?;
+            }
+        }
+        (_, _) => {
+            return Err(internal!(
+                "An unexpected and unhandled stripe event has been received. event: {:?}",
+                event.event_type
+            ))
+        }
+    }
+
+    Ok(())
+}
+
+async fn get_public_key_and_stripe_user_info(
+    event: &WebhookEvent, con: &mut Connection, customer_id: &str,
+) -> Result<(PublicKey, StripeUserInfo), ServerError<StripeWebhookError>> {
+    let public_key: PublicKey = con
+        .maybe_json_get(public_key_from_stripe_customer_id(customer_id))
+        .await?
+        .ok_or(internal!(
+            "There is no public_key related to this customer_id: {:?}",
+            customer_id
+        ))?;
+
+    let user_info: StripeUserInfo = con
+        .maybe_json_get(stripe_user_info(&public_key))
+        .await?
+        .ok_or(internal!(
```

Comment:
The `internal!` macro logs the message, so shouldn't these be `ok_or_else`? Before changing it, I'm curious, have these been logging all the time, even when the value is `Some`?

Reply:
It would be logging all the time, but the thing is, this code is only passed through when a subscription cycle happens.

Reply:
What I mean is I think the message `"There is _no_ ..."` will be logged even if there is `Some`. Just wondering if that's the case, but I just saw it's changed to ok_or_else so :+1: